### PR TITLE
Require an edit cookie to be able to view a comment to be edited.

### LINKED ISSUE #679: Privacy issues
The `/id` endpoint makes it way too easy for someone to download all comments on a site, even ones on pages which aren't from publicly-known URIs. This is a potential privacy concern for folks using isso in a semi-private context, such as comments on friends-only posts, and this is made much worse by the use of auto-incrementing comment IDs (see the recent Parler privacy fiasco for a perfect example of why this is bad).

It looks like the frontend uses this endpoint for the purpose of rendering a submitted comment, but it seems like this could be better done by the `POST` submission API providing the rendered HTML (which would also help with issue #673), or by requiring the authentication cookie to prevent fetching random messages, or even just requiring the thread's page URI as a safety measure.

It's notable that this endpoint is not documented on the [API docs](https://posativ.org/isso/docs/extras/api/) which implies that it's not intended to be known about, but security through obscurity is no security at all.

So, I have two suggestions, both of which should be implemented:

1. Stop using auto-incrementing comment IDs
2. Don't allow arbitrary retrieval based only on the comment ID

### PR DESCRIPTION
**Pull request by fluffy-critter**
The `/id` GET endpoint is only used to retrieve the raw text of a comment to be edited. Being able to arbitrarily retrieve *any* comment through this endpoint is a potential privacy issue, since it allows a malicious actor to individually retrieve all comments from a website without knowledge of an existing thread URI.

This change requires that comment text retrieval has a valid edit cookie.

Fixes #679

### CODE REVIEW

**Comment by fluffy-critter** [comment]
Oops, not quite ready yet - while the unit tests pass, for some reason actual comment submission isn't working now.

File: isso/utils/html.py [hunk 1]
```diff
@@ -76,7 +76,7 @@ class Unofficial(misaka.HtmlRenderer):
 
     def blockcode(self, text, lang):
         lang = ' class="{0}"'.format(html.escape(lang)) if lang else ''
-        return "<pre><code{1}>{0}</code></pre>\n".format(html.escape(text,False), lang)
+        return "<pre><code{1}>{0}</code></pre>\n".format(html.escape(text, False), lang)
```
Comment (ix5) [inline isso/utils/html.py:79]: https://github.com/posativ/isso/pull/680

Comment (fluffy-critter) [reply isso/utils/html.py:79]: Jeeze, of COURSE there's already a PR for someone else making a minor spacing fix. :) FWIW this was put in by autopep8 which I ran to make Travis happy again.

Comment (ix5) [reply isso/utils/html.py:79]: Speaking of travis... another construction site :(
https://www.jeffgeerling.com/blog/2020/travis-cis-new-pricing-plan-threw-wrench-my-open-source-works

Comment (ix5) [reply isso/utils/html.py:79]: Eh, circleci seems like an okay alternative https://app.circleci.com/pipelines/github/ix5/isso/52/workflows/7f941e2b-00f5-44a4-b9c1-cb6b6d2e16d1

Comment (fluffy-critter) [reply isso/utils/html.py:79]: Yeah, that'd be up to @posativ to change, though, I think? My mention of Travis was more just that the pipeline is still running (somehow) and it barfed on the flake8 failure.

File: isso/views/comments.py [hunk 2]
```diff
@@ -351,6 +351,8 @@ def _remote_addr(self, request):
     """
     @api {get} /id/:id view
     @apiGroup Comment
+    @apiDescription
+        View an existing comment, for the purpose of editing. Editing a comment is only possible for a short period of time after it was created and only if the requestor has a valid cookie for it. See the [isso server documentation](https://posativ.org/isso/docs/configuration/server) for details.
```
Comment (ix5) [inline isso/views/comments.py:355]: Maybe change the `apiExample` to
`curl 'https://comments.example.com/id/4' --cokie '4=<cookie>'`

File: isso/views/comments.py [hunk 3]
```diff
@@ -954,6 +960,7 @@ def count(self, environ, request, uri):
     @apiSuccessExample Counts of 5 threads:
         [2, 18, 4, 0, 3]
     """
+
```
Comment (ix5) [inline isso/views/comments.py:963]: Begone!

**Review by ix5 — CHANGES_REQUESTED** [review]
LGTM, but please rebase and remove the now unnecessary formatting commit.

---

I've verified this PR via the following methods:

### Web interface
Creating a comment manually and editing it -> **works**

### curl denied
Creating comment, note id=1, try `curl 'http://localhost/id/1` -> **403** (as expected)

### curl with cookie
Create comment, note id=1, cookie `1=WzIsImUzYjgyMDQwNTY1YmI0YmU2YjExZjc3OGEyZTNkZjMyN2VkMjBhM2IiXQ.YB_Puw.PjiKthPk4XotF-22aV11EoveROw`, try:
```
curl 'http://localhost:8080/id/1' --cookie '1=WzIsIjUwOWU2YzJkNTFlZDRkNjkwNjA5ODcxNzQzODljNzlhNzBlYjQ1NWMiXQ.YB_PKw.GPKGtzNFiHKIwxc3i3U1olTj2jY'
```
->
```
{"id": 1, "parent": null, "created": 1612697387.4656265, "modified": 1612697531.182028, "mode": 1, "text": "<p>Now</p>", "author": "Nameee", "website": "http://web.site", "likes": 0, "dislikes": 0, "notification": 0}
```
As expected, **works**

### nosetests

```
nosetests --with-doctest --with-coverage --cover-package=isso --cover-html isso/
................................................................................
Name                        Stmts   Miss  Cover
-----------------------------------------------
isso/__init__.py              164     57    65%
isso/config.py                 77     11    86%
isso/core.py                   68     20    71%
isso/db/__init__.py            66      6    91%
isso/db/comments.py           135     20    85%
isso/db/preferences.py         17      0   100%
isso/db/spam.py                36      0   100%
isso/db/threads.py             15      1    93%
isso/dispatch.py               39     23    41%
isso/ext/__init__.py           10      0   100%
isso/ext/notifications.py     150    100    33%
isso/migrate.py               201     27    87%
isso/run.py                     5      0   100%
isso/utils/__init__.py         75      8    89%
isso/utils/hash.py             67      3    96%
isso/utils/html.py             50      1    98%
isso/utils/http.py             39     25    36%
isso/utils/parse.py            43      2    95%
isso/views/__init__.py         36      8    78%
isso/views/comments.py        464     70    85%
isso/wsgi.py                  103     23    78%
-----------------------------------------------
TOTAL                        1860    405    78%
----------------------------------------------------------------------
Ran 80 tests in 5.242s

OK
```

File: isso/views/comments.py (hunk 3 above)
Comment (fluffy-critter) [reply isso/views/comments.py:963]: There's quite a few functions with that inconsistent spacing. Also *technically* the docstrings should be inside the functions, not outside of them. Fixing that feels a bit out-of-scope for this change. Maybe in the future there should be a docstring normalization pass.

Comment (fluffy-critter) [reply isso/views/comments.py:963]: (That said, I'll remove that accidentally-introduced newline. No idea how that got in there!)

File: isso/views/comments.py (hunk 2 above)
Comment (fluffy-critter) [reply isso/views/comments.py:355]: Used `-b cookie.txt` to keep it in line with the other `@apiExample` annotations

**Comment by fluffy-critter** [comment]
PR feedback applied.

**Comment by fluffy-critter** [comment]
It'd still be swell if this could get merged in.

**Comment by ix5** [comment]
Hmm, those CI failures do look like a pypi availability issue...

@fluffy-critter maybe rebase again to trigger another CI run?
@jelmer would you mind merging this afterwards?

**Comment by fluffy-critter** [comment]
Looks like the same errors are occurring.

**Comment by ix5** [comment]
The looks like the latest `Flask` package in pypi (version `2.0.2` as of 2021-12-22) only supports python 3.6 and newer.
The latest release to support python 3.5 is https://pypi.org/project/Flask/1.1.4/

Since the [latest PR to get merged](https://github.com/posativ/isso/pull/733) also has the CI runs failing, I'd advocate merging this PR since it introduces no CI regressions.

**Comment by ix5** [comment]
Also, @fluffy-critter you inadvertently (?) slipped some local commits (5 in total) in while rebasing, please get rid of them.

**Comment by fluffy-critter** [comment]
Ah, yeah, totally accidental. Will fix.

File: package.json [hunk 4]
```diff
@@ -4,11 +4,14 @@
   "description": "lightweight Disquis alternative",
   "license": "MIT",
   "repository": "github:posativ/isso",
-  "dependencies": {},
```
Comment (jelmer) [inline package.json:7]: This seems unrelated, or do you need newer versions of these for this change?

File: isso/views/comments.py (hunk 3 above)
Comment (jelmer) [reply isso/views/comments.py:978]: Yeah - I think there's value in fixing some of the formatting, but mixing those changes in with other changes just makes everything harder to review.

**Review by jelmer — CHANGES_REQUESTED** [review]

File: package.json (hunk 4 above)
Comment (fluffy-critter) [reply package.json:7]: I’m not sure how that change snuck in, I thought I’d reverted that file to the one on master.

File: isso/views/comments.py (hunk 3 above)
Comment (fluffy-critter) [reply isso/views/comments.py:978]: A thing I’ve found to be helpful is to make autopep8 part of the build process so that there’s never an opportunity for things to drift in the first place.

Comment (jelmer) [reply isso/views/comments.py:978]: Sure, but that should also be a separate PR :)

**Comment by fluffy-critter** [comment]
Oh, yeah, I absolutely agree.

---
fluffy (she/her)
https://beesbuzz.biz/

**Comment by fluffy-critter** [comment]
PR has been cleaned up and only has PR-specific changes in it now. Sorry about the temporary mess!

**Review by jelmer — APPROVED** [review]
